pennylane_qrack/qrack_device.cpp

```diff
@@ -0,0 +1,791 @@
+
+        if (!is_gpu) {
+            simulatorType.push_back(Qrack::QINTERFACE_CPU);
+        }
+
+        if (is_qbdd) {
+            simulatorType.push_back(Qrack::QINTERFACE_BDT_HYBRID);
+        }
+
+        if (is_hybrid_stabilizer) {
+            simulatorType.push_back(Qrack::QINTERFACE_STABILIZER_HYBRID);
+        }
+
+        if (is_schmidt_decomposed) {
+            simulatorType.push_back(is_schmidt_decomposition_parallel ? Qrack::QINTERFACE_QUNIT_MULTI : Qrack::QINTERFACE_QUNIT);
+        }
+
+        if (is_tensor_network) {
+            simulatorType.push_back(Qrack::QINTERFACE_TENSOR_NETWORK);
+        }
+
+        // (...then reverse:)
+        std::reverse(simulatorType.begin(), simulatorType.end());
+
+        if (!simulatorType.size()) {
+            simulatorType.push_back(Qrack::QINTERFACE_CPU);
+        }
+
+        qsim = CreateQuantumInterface(simulatorType, 0U, Qrack::ZERO_BCI, nullptr, Qrack::CMPLX_DEFAULT_ARG, false, true, is_host_pointer);
+    }
+
+    QrackDevice &operator=(const QuantumDevice &) = delete;
+    QrackDevice(const QrackDevice &) = delete;
+    QrackDevice(QrackDevice &&) = delete;
+    QrackDevice &operator=(QuantumDevice &&) = delete;
+
+    auto AllocateQubit() -> QubitIdType override {
+        return qsim->Allocate(1U);
+    }
+    auto AllocateQubits(size_t num_qubits) -> std::vector<QubitIdType> override {
+        std::vector<QubitIdType> ids(num_qubits);
+        for (size_t i = 0U; i < num_qubits; ++i) {
+            ids[i] = qsim->Allocate(1U);
+        }
+        return ids;
+    }
+    [[nodiscard]] auto Zero() const -> Result override { return const_cast<Result>(&QRACK_RESULT_FALSE_CONST); }
+    [[nodiscard]] auto One() const -> Result override { return const_cast<Result>(&QRACK_RESULT_TRUE_CONST); }
+    auto Observable(ObsId id, const std::vector<std::complex<double>> &matrix,
+                    const std::vector<QubitIdType> &wires) -> ObsIdType override
+    {
+        // if (id == ObsId::Hermitian) {
+        //     return obs_manager.createHermitianObs(matrix, wires);
+        // }
+        //
+        // return obs_manager.createNamedObs(id, wires);
+        return 0;
+    }
+    auto TensorObservable(const std::vector<ObsIdType> &obs) -> ObsIdType override
+    {
+        // return obs_manager.createTensorProdObs(obs);
+        return 0;
+    }
+    auto HamiltonianObservable(const std::vector<double> &coeffs, const std::vector<ObsIdType> &obs)
+        -> ObsIdType override
+    {
+        // return obs_manager.createHamiltonianObs(coeffs, obs);
+        return 0;
+    }
+    auto Measure(QubitIdType id, std::optional<int> postselect) -> Result override {
+        bool *ret = (bool *)malloc(sizeof(bool));
+        if (postselect.has_value()) {
+            *ret = qsim->ForceM(id, postselect.value());
+        } else {
+            *ret = qsim->M(id);
+        }
+        return ret;
+    }
+
+    void ReleaseQubit(QubitIdType id) override
```

Comment:
[nitpick]: Stylistically, it seems sometimes functions abide by camelcase and others by capitalizing the first letter. Might be a good reason why, but wanted to flag.

Reply:
Specifically, public methods are Pascal case and (implicitly) "private" methods are camel case. Since there is a standard convention, I'm going to consider this comment superseded. (Sorry.) PennyLane does things like mixing camel case and snake case in their API, and we're basically just lifting method prototypes from them. C++ should not adopt Python standards, but we should follow a style guide like the Catalyst developers themselves.